controller/kubernetes_controller.go

```diff
@@ -0,0 +1,409 @@
+	if pv.Spec.ClaimRef != nil && pv.Status.Phase == v1.VolumeBound && ks.PVStatus != string(pv.Status.Phase) {
+		ks.PVCName = pv.Spec.ClaimRef.Name
+		ks.Namespace = pv.Spec.ClaimRef.Namespace
+		ks.LastPVCRefAt = ""
+	}
+
+	// PVC is no longer bound with PV. indicating history data by setting <LastPVCRefAt>
+	if ks.PVStatus == string(v1.VolumeBound) && ks.PVStatus != string(pv.Status.Phase) {
+		ks.LastPVCRefAt = util.Now()
+	}
+
+	var pod *v1.Pod
+	pod, err = kc.getMountPod(ks.Namespace, ks.PVCName)
+	if err != nil {
+		return err
+	}
+
+	// pod deleted. indicating history data by setting <LastPodRefAt>
+	if ks.PodName != "" && ks.LastPodRefAt == "" && (pod == nil || pod.DeletionTimestamp != nil) {
+		if ks.LastPodRefAt == "" {
+			ks.LastPodRefAt = util.Now()
+		}
+	}
+	if pod != nil {
+		workloadName, workloadType := kc.detectWorkload(pod, ks)
+		ks.PodName = pod.Name
+		ks.PodStatus = string(pod.Status.Phase)
+		ks.WorkloadName = workloadName
+		ks.WorkloadType = workloadType
+		ks.LastPodRefAt = ""
+	}
+
+	// if pv.Status.Phase == v1.VolumePending || pv.Status.Phase == v1.VolumeAvailable, do nothing
+	ks.PVName = name
+	ks.PVStatus = string(pv.Status.Phase)
+
+	volume, err = kc.ds.UpdateVolume(volume)
+	if err != nil {
+		return err
+	}
+	kc.eventRecorder.Eventf(volume, v1.EventTypeNormal, EventReasonStart, "Updating KubernetesStatus for volume %v", volume.Name)
+
+	return nil
+}
+
+func IsErrorNotFound(err error) bool {
+	return err != nil && strings.HasSuffix(strings.Trim(err.Error(), "\n"), "not found")
+}
+
+func (kc *KubernetesController) validatePersistentVolume(pv *v1.PersistentVolume) (string, error) {
+	if pv == nil {
+		return "", nil
+	}
+	if pv.Spec.FlexVolume != nil {
+		return "", nil
+	}
+	// try to get associated Longhorn volume name
+	if pv.Spec.CSI == nil || pv.Spec.CSI.VolumeHandle == "" {
+		return "", nil
+	}
+	return pv.Spec.CSI.VolumeHandle, nil
+}
+
+func (kc *KubernetesController) enqueuePersistentVolume(pv *v1.PersistentVolume) {
+	key, err := controller.KeyFunc(pv)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("Couldn't get key for object %#v: %v", pv, err))
+		return
+	}
+	kc.queue.AddRateLimited(key)
+	return
+}
+
+func (kc *KubernetesController) enqueuePodChange(pod *v1.Pod) {
+	if _, err := controller.KeyFunc(pod); err != nil {
+		utilruntime.HandleError(fmt.Errorf("Couldn't get key for object %#v: %v", pod, err))
+		return
+	}
+
+	for _, v := range pod.Spec.Volumes {
+		if v.VolumeSource.PersistentVolumeClaim != nil {
+			pvc, err := kc.pvcLister.PersistentVolumeClaims(pod.Namespace).Get(v.VolumeSource.PersistentVolumeClaim.ClaimName)
+			if IsErrorNotFound(err) {
+				continue
+			}
+			if err != nil {
+				utilruntime.HandleError(fmt.Errorf("Couldn't get key for object %#v: %v", pvc, err))
+				return
+			}
+			pvName := pvc.Spec.VolumeName
+			if pvName != "" {
+				kc.queue.AddRateLimited(pvName)
+			}
+		}
+	}
+	return
+}
+
+func (kc *KubernetesController) getFromVCache(pvName string) string {
```

Comment:
Just `getxxx`/`setxxx`/`delete`

In fact, I just found https://golang.org/pkg/sync/#Map can do the job natively.